src/error.rs

```diff
@@ -257,11 +341,11 @@ pub enum ErrorKind {
 
     /// Wrapper around `bson::de::Error`.
     #[error("{0}")]
-    BsonDecode(crate::bson::de::Error),
+    BsonDeserializationError(crate::bson::de::Error),
```

Comment:
I updated the terminology here and appended the "Error" suffix to the end to match the other cases. I wonder if it would actually be better if we dropped all the "Error" suffixes though, since they're kind of redundant with `ErrorKind`. I'm not really sure which way would be better, but I just opted for this since it was the least amount of changes.
```rust
match *e.kind {
    ErrorKind::IoError(e) => { ... },
    ErrorKind::CommandError(e) => { ... },
    ErrorKind::ArgumentError { message } => { ... },
    // rest of cases...
}

// vs
match *e.kind {
    ErrorKind::Io(e) => { ... },
    ErrorKind::Command(e) => { ... },
    ErrorKind::Argument { message } => { ... },
    // rest of cases...
}
```

Reply:
There's actually a [relevant clippy lint](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/enum_variants.rs) here that complains when all enum variants have the same suffix; I ran into this once before when I was making the `Retryability` enum and all of the variants were suffixed with `Retryable` (i.e. `ReadRetryable`, `WriteRetryable`, `NonRetryable`). I think we're avoiding it for now since `SessionsNotSupported` currently doesn't have the `Error` suffix as you point out below, but I do think it would be nicer to get rid of the `Error` in the names in the spirit of this lint. I don't feel super strongly about this though so I'll leave it up to you; I agree it's best to minimize changes when we don't have a strong opinion on something.

Reply:
given that, dropping `Error` SGTM

Reply:
Updated to drop suffix. I added the `Invalid` prefix to the `Argument`, `Response`, and `TlsConfig` variants to convey that something malformed was produced somewhere.